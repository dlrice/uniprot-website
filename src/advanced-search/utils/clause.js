import { v1 } from 'uuid';
import { serializableDeepAreEqual, removeProperty } from '../../utils/utils';

export const createEmptyClause = () => ({
  id: v1(),
  logicOperator: 'AND',
  field: {
    label: 'Any',
    term: 'All',
    example: 'a4_human, P05067, cdc7 human',
    itemType: 'single',
    dataType: 'string',
    id: 'id_Any',
  },
  queryInput: {},
});

// export const isClauseTouched = (clause) => {
//   const empty = createEmptyClause();
//   const emptyWithoutID = removeProperty(empty, 'id');
//   const clauseWithoutID = removeProperty(clause, 'id');
//   return !serializableDeepEquals(emptyWithoutID, clauseWithoutID);
// };

export const clausesAreEqual = (clause1, clause2) => serializableDeepAreEqual(removeProperty(clause1, 'id'), removeProperty(clause2, 'id'));
