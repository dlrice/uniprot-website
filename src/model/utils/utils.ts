/* eslint-disable @typescript-eslint/no-explicit-any */
const hasContent = (obj: any) => {
  return Object.values(obj).some(val => {
    if (Array.isArray(val)) {
      const valArray = val as any[];
      return valArray.length > 0;
    }
    if (typeof val === 'object' && val) {
      return Object.values(val).length > 0;
    }
    return false;
  });
};

export default hasContent;
