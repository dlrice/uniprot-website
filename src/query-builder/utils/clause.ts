import { v1 } from 'uuid';

import {
  Operator,
  Clause,
  ItemType,
  DataType,
  FieldType,
} from '../types/searchTypes';

export const createEmptyClause = (): Clause => ({
  id: v1(),
  logicOperator: Operator.AND,
  searchTerm: {
    label: 'All',
    term: 'All',
    example: 'a4_human, P05067, cdc7 human',
    itemType: ItemType.single,
    dataType: DataType.string,
    fieldType: FieldType.general,
    id: 'id_all',
  },
  queryBits: {},
});

export const createPreSelectedClauses = (): Clause[] => [
  {
    id: v1(),
    logicOperator: Operator.AND,
    searchTerm: {
      label: 'Gene Name [GN]',
      term: 'gene',
      example: 'ydj1',
      itemType: ItemType.single,
      dataType: DataType.string,
      fieldType: FieldType.general,
      id: 'gene_field',
    },
    queryBits: {},
  },
  {
    id: v1(),
    logicOperator: Operator.AND,
    searchTerm: {
      id: 'taxonomy_name',
      label: 'Taxonomy [OC]',
      itemType: ItemType.single,
      term: 'taxonomy_name',
      dataType: DataType.string,
      fieldType: FieldType.general,
      example: 'human',
      autoComplete: '/uniprot/api/suggester?dict=taxonomy&query=?',
      autoCompleteQueryTerm: 'taxonomy_id',
    },
    queryBits: {},
  },
  {
    id: v1(),
    logicOperator: Operator.AND,
    searchTerm: {
      id: 'keyword_field',
      label: 'Keyword [KW]',
      term: 'keyword',
      example: 'chromosomal',
      itemType: ItemType.single,
      dataType: DataType.string,
      fieldType: FieldType.general,
      autoComplete: '/uniprot/api/suggester?dict=keyword&query=?',
      autoCompleteQueryTerm: 'keyword',
    },
    queryBits: {},
  },
  createEmptyClause(),
];
