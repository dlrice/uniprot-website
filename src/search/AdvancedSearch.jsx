import React from 'react';
import ClauseList from './ClauseList';

const AdvancedSearch = (props) => {
  const { handleAdvancedSubmitClick, namespace, dispatchAddClause } = props;
  return (
    <div className="advanced-search">
      <div>
        <label htmlFor="namespace-select">
          Searching in
          <select id="namespace-select">
            <option>{namespace}</option>
          </select>
        </label>
      </div>
      <ClauseList {...props} />
      <button type="button" id="add-field" className="button" onClick={dispatchAddClause}>
        Add Field
      </button>
      <div>
        <button
          type="button"
          id="submit-query"
          className="button"
          onClick={handleAdvancedSubmitClick}
        >
          Search
        </button>
      </div>
    </div>
  );
};

export default AdvancedSearch;
