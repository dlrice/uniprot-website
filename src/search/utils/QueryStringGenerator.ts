import { Clause } from '../types/searchTypes';

type IPrefixMap = {
  feature: string;
  comment: string;
  [key: string]: string;
};

const getItemTypePrefix = (itemType: string) => {
  const itemTypeToPrefixMap: IPrefixMap = {
    feature: 'ft_',
    comment: 'cc_',
  };
  return itemTypeToPrefixMap[itemType] || '';
};

const getItemTypeEvidencePrefix = (itemType: string) => {
  const itemTypeToEvidencePrefixMap: IPrefixMap = {
    feature: 'ftev_',
    comment: 'ccev_',
  };
  return itemTypeToEvidencePrefixMap[itemType] || '';
};

const getItemTypeRangePrefix = (itemType: string) =>
  itemType === 'feature' ? 'ftlen_' : '';

const createTermString = (
  term: string | undefined,
  itemType: string,
  id: string | undefined
) => {
  if (term === undefined) {
    throw new Error('term is undefined');
  }
  if (term === 'ec') {
    if (id) {
      return 'ec:';
    }
    throw new Error('ID value not provided in query');
  }
  if (['organism', 'taxonomy', 'host'].includes(term)) {
    if (id) {
      return `${term}_id:`;
    }
    return `${term}_name:`;
  }
  const itemTypePrefix = getItemTypePrefix(itemType);
  return `${itemTypePrefix}${term}${term ? ':' : ''}`;
};

const createValueString = (
  term: string | undefined,
  valuePrefix: string | undefined,
  stringValue: string,
  id: string | undefined
) => {
  if (term === undefined) {
    throw new Error('term is undefined');
  }
  if (term === 'ec') {
    if (id) {
      return id;
    }
    throw new Error('ID value not provided in query');
  }
  if (id) {
    if (['cofactor_chebi', 'catalytic_activity'].includes(term)) {
      return `"${id}"`;
    }
    if (
      ['organism', 'taxonomy', 'host', 'keyword', 'scl_term'].includes(term)
    ) {
      return id;
    }
  }

  // We are testing for term=xref and valuePrefix=any because the
  // search API expects the valuePrefix to be ommited in this case.
  // eg xref:foo rather than xref_any:foo
  const valuePrefixChecked =
    valuePrefix && !(term === 'xref' && valuePrefix === 'any')
      ? `${valuePrefix}-`
      : '';
  return `${valuePrefixChecked}${stringValue}`;
};

const createSimpleSubquery = (clause: Clause) => {
  if (clause.searchTerm.itemType === 'group') {
    throw Error('Cannot create a query with a group term.');
  }
  const { itemType, term, valuePrefix } = clause.searchTerm;
  const { stringValue, id } = clause.queryInput;
  if (!stringValue) {
    throw new Error('Value not provided in query');
  }
  if (term === 'All') {
    return stringValue;
  }
  const termString = createTermString(term, itemType, id);
  const valueString = createValueString(term, valuePrefix, stringValue, id);
  return `(${termString}${valueString})`;
};

const createRangeSubquery = (clause: Clause) => {
  const { term, itemType } = clause.searchTerm;
  const { rangeFrom, rangeTo } = clause.queryInput;
  const rangeFromChecked = rangeFrom || '';
  const rangeToChecked = rangeTo || '';
  const itemTypeRangePrefix = getItemTypeRangePrefix(itemType);
  return `(${itemTypeRangePrefix}${term}:[${rangeFromChecked} TO ${rangeToChecked}])`;
};

const getEvidenceSubquery = (clause: Clause) => {
  const { evidenceValue } = clause.queryInput;
  const { term, itemType } = clause.searchTerm;
  if (!evidenceValue) {
    throw new Error('Evidence value not provided');
  }
  const itemTypeEvidencePrefix = getItemTypeEvidencePrefix(itemType);
  return `(${itemTypeEvidencePrefix}${term}:${evidenceValue})`;
};

const createQueryString = (clauses: Clause[] = []): string =>
  clauses.reduce((queryAccumulator: string, clause: Clause) => {
    const query = [];
    if (
      (clause.queryInput.id && clause.queryInput.id !== '') ||
      (clause.queryInput.stringValue && clause.queryInput.stringValue !== '')
    ) {
      query.push(createSimpleSubquery(clause));
    }
    if (clause.queryInput.rangeFrom || clause.queryInput.rangeTo) {
      query.push(createRangeSubquery(clause));
    }
    if (
      clause.queryInput.evidenceValue &&
      clause.queryInput.evidenceValue !== ''
    ) {
      query.push(getEvidenceSubquery(clause));
    }
    let queryJoined = query.join(' AND ');
    if (query.length > 1) {
      queryJoined = `(${queryJoined})`;
    }

    let logicOperator = '';
    if (queryAccumulator.length > 0 && query.length > 0) {
      logicOperator = ` ${clause.logicOperator} `;
    } else if (
      queryAccumulator.length === 0 &&
      clause.logicOperator === 'NOT'
    ) {
      logicOperator = `${clause.logicOperator} `;
    }

    return `${queryAccumulator}${logicOperator}${queryJoined}`;
  }, '');

export default createQueryString;
