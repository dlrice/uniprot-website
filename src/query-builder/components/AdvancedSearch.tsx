import React, { FC, FormEvent, useState, useEffect } from 'react';
import { useHistory, useRouteMatch, generatePath } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import qs from 'query-string';
import { PageIntro } from 'franklin-sites';

import ClauseList from './ClauseList';

import useDataApi from '../../shared/hooks/useDataApi';

import { createEmptyClause, createPreSelectedClauses } from '../utils/clause';
import { stringify, parse } from '../utils/queryStringProcessor';

import { addMessage } from '../../messages/state/messagesActions';

import apiUrls from '../../shared/config/apiUrls';
import { Namespace, NamespaceLabels } from '../../shared/types/namespaces';
import { LocationToPath, Location } from '../../app/config/urls';

import {
  MessageFormat,
  MessageLevel,
} from '../../messages/types/messagesTypes';
import { Clause, SearchTermType } from '../types/searchTypes';

import '../../uniprotkb/components/search/styles/search-container.scss';
import './styles/advanced-search.scss';

const flatten = (searchTermData: SearchTermType[]): SearchTermType[] => {
  return searchTermData.flatMap((searchTermDatum: SearchTermType) => {
    if (searchTermDatum.siblings) {
      return flatten(searchTermDatum.siblings);
    }
    if (searchTermDatum.items) {
      return flatten(searchTermDatum.items);
    }
    return searchTermDatum;
  });
};

const AdvancedSearch: FC = () => {
  const history = useHistory();
  const match = useRouteMatch<{ namespace?: Namespace }>(
    LocationToPath[Location.QueryBuilder]
  );
  const dispatch = useDispatch();

  // To be replaced by getting it from url
  const [clauses, setClauses] = useState<Clause[]>([]);

  const namespace = match?.params?.namespace;

  const { data: searchTermsData } = useDataApi<SearchTermType[]>(
    namespace && apiUrls.advancedSearchTerms(namespace)
  );

  // if URL doesn't finish with a namespace redirect to "uniprotkb" by default
  useEffect(() => {
    if (!namespace) {
      history.replace(
        history.createHref({
          ...history.location,
          pathname: generatePath(LocationToPath[Location.QueryBuilder], {
            namespace: Namespace.uniprotkb,
          }),
        })
      );
    }
  }, [history, namespace]);

  useEffect(() => {
    if (!searchTermsData) {
      return;
    }
    const query = qs.parse(history.location.search, { decode: true })?.query;
    const parsedQuery =
      query && !Array.isArray(query) ? parse(query) : undefined;

    // flatten all the endpoint-described clauses to be able to to look-up
    const flattened = flatten(searchTermsData);
    // for each parsed clause, try to find the corresponding endpoint-described
    // clause to merge its 'searchTerm' field
    const validatedQuery: Clause[] = [];
    for (const clause of parsedQuery || []) {
      const found = flattened.find(
        (item) => item.term === clause.searchTerm.term
      );
      // if it exists, assign it 'searchTerm'
      if (found) {
        validatedQuery.push({ ...clause, searchTerm: found });
      } else {
        dispatch(
          addMessage({
            id: clause.searchTerm.term,
            content: `"${clause.searchTerm.term}" is not a valid query term for ${namespace}`,
            format: MessageFormat.POP_UP,
            level: MessageLevel.FAILURE,
          })
        );
      }
    }

    setClauses((clauses) => {
      if (clauses.length) {
        return clauses;
      }
      if (validatedQuery.length) {
        return validatedQuery;
      }
      return createPreSelectedClauses();
    });
  }, [dispatch, history.location.search, namespace, searchTermsData]);

  if (!searchTermsData || !namespace) {
    return null;
  }

  const addClause = () => {
    setClauses((clauses) => [...clauses, createEmptyClause()]);
  };

  const removeClause = (clauseId: string) => {
    setClauses((clauses) => {
      if (clauses.length === 1) {
        return [createEmptyClause()];
      }
      return clauses.filter((clause) => clause.id !== clauseId);
    });
  };

  const handleSubmit = (event: FormEvent) => {
    event.preventDefault();
    const queryString = stringify(clauses);
    history.push({
      pathname: `/${namespace}`,
      search: queryString && `query=${queryString}`,
    });
  };

  return (
    <>
      <PageIntro title="Advanced search" />

      <form
        className="advanced-search"
        onSubmit={handleSubmit}
        data-testid="advanced-search-form"
      >
        <fieldset>
          <label htmlFor="namespace-select">
            Searching in
            <select
              id="namespace-select"
              onChange={(e) => {
                history.replace(
                  history.createHref({
                    pathname: generatePath(
                      LocationToPath[Location.QueryBuilder],
                      { namespace: e.target.value }
                    ),
                  })
                );
              }}
              value={namespace}
            >
              {Object.keys(NamespaceLabels).map((key) => (
                <option value={key} key={key}>
                  {NamespaceLabels[key as Namespace]}
                </option>
              ))}
            </select>
          </label>
        </fieldset>
        <fieldset>
          <ClauseList
            removeClause={removeClause}
            clauses={clauses}
            setClauses={setClauses}
            searchTerms={searchTermsData}
          />
        </fieldset>
        <div className="advanced-search__actions">
          <button
            type="button"
            id="add-field"
            className="button tertiary"
            data-testid="advanced-search-add-field"
            onClick={addClause}
          >
            Add Field
          </button>
          <button type="submit" id="submit-query" className="button">
            Search
          </button>
        </div>
      </form>
    </>
  );
};

export default AdvancedSearch;
