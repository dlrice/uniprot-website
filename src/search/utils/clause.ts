import { v1 } from 'uuid';
import { serializableDeepAreEqual, removeProperty } from '../../utils/utils';
import { Operator, Input, SearchTermType } from '../types/searchTypes';

enum itemType {
  comment = 'comment',
  database = 'database',
  feature = 'feature',
  goterm = 'goterm',
  group = 'group',
  groupDisplay = 'groupDisplay',
  single = 'single',
}

enum dataType {
  empty = '',
  date = 'date',
  enum = 'enum',
  integer = 'integer',
  string = 'string',
}

type Clause = {
  id: string;
  logicOperator: Operator;
  searchTerm: {
    label: string;
    term: string;
    example: string;
    itemType: itemType;
    dataType: dataType;
    id: string;
  };
  queryInput: Input;
};

export const createEmptyClause = (): Clause => ({
  id: v1(),
  logicOperator: Operator.AND,
  searchTerm: {
    label: 'Any',
    term: 'All',
    example: 'a4_human, P05067, cdc7 human',
    itemType: itemType.single,
    dataType: dataType.string,
    id: 'id_Any',
  },
  queryInput: {},
});

export const clausesAreEqual = (clause1: Clause, clause2: Clause) =>
  serializableDeepAreEqual(
    removeProperty(clause1, 'id'),
    removeProperty(clause2, 'id')
  );

const findSearchTermRecursive = (
  queryField: string,
  searchTerms: SearchTermType[],
  valuePrefix = ''
): SearchTermType | undefined => {
  for (const searchTerm of searchTerms) {
    if (
      valuePrefix &&
      searchTerm.valuePrefix === valuePrefix &&
      searchTerm.term === queryField
    ) {
      return searchTerm;
    }
    if (!valuePrefix && searchTerm.term === queryField) {
      return searchTerm;
    }
    if (searchTerm.items) {
      const found = findSearchTermRecursive(
        queryField,
        searchTerm.items,
        valuePrefix
      );
      if (found) {
        return found;
      }
    }
  }
};

const findSearchTerm = (
  queryField: string,
  searchTerms: SearchTermType[],
  valuePrefix = ''
): SearchTermType | undefined => {
  const searchTerm = findSearchTermRecursive(
    queryField,
    searchTerms,
    valuePrefix
  );
  if (!searchTerm) {
    throw new Error(`${queryField} not a valid field.`);
  }
  return searchTerm;
};

const parseRangeValue = (value: string) => {
  const regeNumberRange = /\[([0-9]+)\s*TO\s*([0-9]+)\]/gi;
  const captureGroupsNumber = regeNumberRange.exec(value);
  if (captureGroupsNumber && captureGroupsNumber[1] && captureGroupsNumber[2]) {
    return {
      rangeFrom: Number(captureGroupsNumber[1]),
      rangeTo: Number(captureGroupsNumber[2]),
    };
  }
  const regexDateRange = /\[([0-9\-]+)\s*TO\s*([0-9\-]+)\]/gi;
  const captureGroupsDate = regexDateRange.exec(value);
  if (captureGroupsDate && captureGroupsDate[1] && captureGroupsDate[2]) {
    return {
      rangeFrom: captureGroupsDate[1],
      rangeTo: captureGroupsDate[2],
    };
  }
  throw new Error(`${value} value is not a valid range`);
};

const parseXrefSubquery = (
  term: string,
  value: string,
  conjunction: keyof typeof Operator,
  searchTerms: SearchTermType[]
) => {
  const clause = {
    id: v1(),
    logicOperator: conjunction,
  };
  const tokens = value.split('-');
  if (tokens.length === 1) {
    // Only a value is provided so this is an xref any search
    return {
      ...clause,
      searchTerm: findSearchTerm('xref', searchTerms, 'any'),
      queryInput: { stringValue: tokens[0] },
    } as Clause;
  }
  if (tokens.length === 2) {
    return {
      ...clause,
      searchTerm: findSearchTerm('xref', searchTerms, tokens[0]),
      queryInput: { stringValue: tokens[1] },
    } as Clause;
  }
  throw new Error(`${value} is not a properly formed xref.`);
};

const parseIdNameSubquery = (
  queryTerm: string,
  value: string,
  conjunction: keyof typeof Operator,
  searchTerms: SearchTermType[]
) => {
  const [termName, searchType] = queryTerm.split('_');
  const searchTerm = findSearchTerm(termName, searchTerms);

  if (searchType === 'id') {
    return {
      id: v1(),
      logicOperator: conjunction,
      searchTerm,
      queryInput: { stringValue: value },
    } as Clause;
  }
  throw new Error(`${queryTerm} not a valid _id or _name style term.`);
};

const parseRangeOrEvidenceSubquery = (subquery: string, value: string) => {
  const regexp = /^(?:ft|cc)(len|ev)?_(.*):.*/i;
  const matches = regexp.exec(subquery);
  if (!matches) {
    return {};
  }
  const rangeOrEvidence = matches[1];
  if (!rangeOrEvidence) {
    return { stringValue: value };
  }
  if (rangeOrEvidence === 'ev') {
    return { evidenceValue: value };
  }
  if (rangeOrEvidence === 'len') {
    return { ...parseRangeValue(value) };
  }
  throw new Error(`${subquery} subquery is not valid.`);
};

const parseSinglePartSubquery = (
  conjunctionString: string = 'AND',
  termValue: string,
  searchTerms: SearchTermType[]
) => {
  if (!searchTerms) {
    return;
  }
  const conjunctionUpper = conjunctionString.toUpperCase();

  let conjunction: keyof typeof Operator;
  if (conjunctionUpper in Operator) {
    conjunction = conjunctionUpper as keyof typeof Operator;
  } else {
    throw new Error(`${conjunctionString} conjunction is not valid.`);
  }

  const tokens = termValue.split(':');
  const value = tokens[1];
  let term = tokens[0];
  if (!value) {
    const emptyClause = createEmptyClause();
    return { ...emptyClause, queryInput: { stringValue: term } };
  }

  if (term.match(/^xref/i)) {
    return parseXrefSubquery(term, value, conjunction, searchTerms);
  }

  if (term.match(/^organism|taxonomy/i)) {
    return parseIdNameSubquery(term, value, conjunction, searchTerms);
  }

  let queryInput = {};
  if (term.match(/^ft(len|ev)?_/i)) {
    try {
      queryInput = parseRangeOrEvidenceSubquery(termValue, value);
    } catch (e) {
      // console.log(e);
    }
    term = getSinglePartSubqueryTerm(termValue);
  }

  const searchTerm = findSearchTerm(term, searchTerms);
  if (!searchTerm) {
    throw new Error(`${term} not a valid term.`);
  }

  return {
    id: v1(),
    logicOperator: conjunction,
    searchTerm,
    queryInput: {
      ...queryInput,
      ...(searchTerm.hasRange
        ? parseRangeValue(value)
        : { stringValue: value }),
    },
  } as Clause;
};

interface IPrefixMap {
  ft: string;
  cc: string;
  [key: string]: string;
}

const itemTypeToPrefixMap: IPrefixMap = {
  ft: 'feature',
  cc: 'comment',
};

const getSinglePartSubqueryPrefix = (subquery: string) => {
  const prefix = subquery.slice(0, 2);
  return prefix in itemTypeToPrefixMap ? prefix : null;
};

const allStringArrayElementsEqual = (arr: Array<string | null>) => {
  return arr.every(
    (element: string | null, i: number) => !!element && element === arr[0]
  );
};

const getTermFrom = (queryStringGroup: string[]) => {
  const prefixes = queryStringGroup.map((queryString: string) =>
    queryString.slice(0, 2)
  );
  if (prefixes.every((prefix: string, i: number) => prefix === prefixes[0])) {
    return itemTypeToPrefixMap[prefixes[0]];
  } else {
    throw new Error(`${queryStringGroup} has inconsistent item type prefixes.`);
  }
};

const getSinglePartSubqueryTerm = (field: string) => {
  if (field.indexOf('_') < 0) {
    return field;
  }
  const regexp = /^\w*?_(\w+)/g;
  const match = regexp.exec(field);
  if (match && match[1]) {
    return match[1];
  }
  throw new Error(`${field} does not contain a searchtype.`);
};

const getMultiplePartSubqueryTerm = (subqueryParts: string[]) => {
  const terms = subqueryParts.map(subqueryPart =>
    getSinglePartSubqueryTerm(subqueryPart)
  );
  if (allStringArrayElementsEqual(terms)) {
    return terms[0];
  }
};

const getMultiplePartSubqueryPrefix = (subqueryParts: string[]) => {
  const prefixes = subqueryParts.map(subqueryPart =>
    getSinglePartSubqueryPrefix(subqueryPart)
  );
  if (allStringArrayElementsEqual(prefixes)) {
    return prefixes[0];
  }
};

const parseMultiplePartSubquery = (
  conjunction: string,
  subqueryParts: string[],
  searchTerms: SearchTermType[]
) => {
  const term = getMultiplePartSubqueryTerm(subqueryParts);
  if (!term) {
    throw new Error(`${subqueryParts} has inconsistent terms.`);
  }
  const searchTerm = findSearchTerm(term, searchTerms);
  if (!searchTerm) {
    throw new Error(`Cannot find ${term}: in search_terms.`);
  }
  const prefix = getMultiplePartSubqueryPrefix(subqueryParts);
  if (!prefix) {
    throw new Error(`${subqueryParts} has inconsistent prefixes.`);
  }
  let queryInput = {};
  subqueryParts.forEach((subqueryPart: string) => {
    const [field, value] = subqueryPart.split(':');
    queryInput = {
      ...queryInput,
      ...parseRangeOrEvidenceSubquery(subqueryPart, value),
    };
  });

  return {
    id: v1(),
    logicOperator: conjunction,
    searchTerm,
    queryInput,
  };
};

const parseSubquery = (
  conjunction: string,
  subquery: string,
  searchTerms: SearchTermType[]
) => {
  const regexQuery = /\(?([0-9a-z_]+:?[0-9a-z\[\]\s\-_\.]*)\)?(?:\s*(AND|OR|NOT)\s*)*/gi;
  const subqueryParts = [];
  let match;
  do {
    // Need this so we don't assign within a loop: while ((match = regexQuery.exec(subquery)) !== null)
    match = regexQuery.exec(subquery);
    if (match) {
      const fieldValue = match[1];
      subqueryParts.push(fieldValue);
    }
  } while (match);

  switch (subqueryParts.length) {
    case 0:
      throw new Error(`No clauses found in ${subquery}.`);
    case 1:
      return parseSinglePartSubquery(
        conjunction,
        subqueryParts[0],
        searchTerms
      );
    default:
      return parseMultiplePartSubquery(conjunction, subqueryParts, searchTerms);
  }
};

const getTopLevelConjunctionsClauses = (query: string) => {
  let balance = 0;
  const clauses = [];
  const conjunctions = [];
  let start = 0;
  for (let i = 0; i < query.length; i += 1) {
    const character = query[i];
    if (character === '(') {
      balance += 1;
      if (balance === 1) {
        if (i === 0) {
          conjunctions.push('AND');
        } else {
          conjunctions.push(query.slice(start + 1, i).trim());
        }
        start = i;
      }
    } else if (character === ')') {
      balance -= 1;
      if (balance === 0) {
        clauses.push(query.slice(start + 1, i));
        start = i;
      }
    }
  }
  return [conjunctions, clauses];
};

export const parseQueryString = (
  query: string,
  searchTerms: SearchTermType[]
) => {
  if (!query) {
    return [createEmptyClause()];
  }
  const [
    topLevelConjunctions,
    topLevelSubqueries,
  ] = getTopLevelConjunctionsClauses(query);
  const clauses = [];
  for (let i = 0; i < topLevelConjunctions.length; i += 1) {
    const conjuction = topLevelConjunctions[i];
    const clause = topLevelSubqueries[i];
    clauses.push(parseSubquery(conjuction, clause, searchTerms));
  }
  return clauses;
};
