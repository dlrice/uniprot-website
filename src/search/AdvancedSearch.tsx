import React, { Component } from 'react';
import ClauseList from './ClauseList';
import { unpackQueryUrl } from '../utils/apiUrls';
import { Namespace } from './types/searchTypes';
import './styles/AdvancedSearch.scss';

type AdvancedSearchProps = {
  handleAdvancedSubmitClick: () => void;
  namespace: Namespace;
  dispatchAddClause: () => void;
};
class AdvancedSearch extends Component<AdvancedSearchProps> {
  componentDidMount() {
    const {
      searchTerms,
      dispatchUpdateClauses,
      dispatchfetchEvidencesIfNeeded,
      dispatchFetchSearchTermsIfNeeded,
      queryString,
    } = this.props;
    dispatchfetchEvidencesIfNeeded('go');
    dispatchfetchEvidencesIfNeeded('annotation');
    dispatchFetchSearchTermsIfNeeded();

    if (searchTerms.length && queryString) {
      dispatchUpdateClauses(unpackQueryUrl(queryString, searchTerms));
    }
  }

  componentDidUpdate(prevProps) {
    const { queryString: prevQueryString, searchTerms: prevSearchTerms } = prevProps;
    const { queryString, searchTerms, dispatchUpdateClauses } = this.props;
    if (
      searchTerms.length
      && queryString
      && (queryString !== prevQueryString || searchTerms !== prevSearchTerms)
    ) {
      dispatchUpdateClauses(unpackQueryUrl(queryString, searchTerms));
    }
  }

  render() {
    const { handleAdvancedSubmitClick, namespace, dispatchAddClause } = this.props;
    return (
      <div className="advanced-search">
        <div>
          <label htmlFor="namespace-select">
            Searching in
            <select id="namespace-select">
              <option>{namespace}</option>
            </select>
          </label>
        </div>
        <ClauseList {...this.props} />
        <button type="button" id="add-field" onClick={dispatchAddClause}>
          Add Field
        </button>
        <div>
          <button
            type="button"
            id="submit-query"
            className="button"
            onClick={handleAdvancedSubmitClick}
          >
            Search
          </button>
        </div>
      </div>
    );
  }
}

export default AdvancedSearch;
