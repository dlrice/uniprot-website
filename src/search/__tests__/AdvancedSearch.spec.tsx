import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { createEmptyClause } from '../utils/clause';
import AdvancedSearch from '../AdvancedSearch';

configure({ adapter: new Adapter() });

let wrapper;
let props;

describe('AdvancedSearch shallow components', () => {
  beforeEach(() => {
    props = {
      dispatchAddClause: jest.fn(),
      handleAdvancedSubmitClick: jest.fn(),
      dispatchFetchSearchTermsIfNeeded: jest.fn(),
      dispatchfetchEvidencesIfNeeded: jest.fn(),
      dispatchCopyQueryClausesToSearch: jest.fn(),
      history: {
        push: jest.fn(),
      },
      clauses: [...Array(4)].map(() => createEmptyClause()),
      namespace: 'UniProtKB',
      searchTerms: {
        data: [],
      },
      evidences: {
        go: {
          data: [],
          isFetching: false,
        },
        annotation: {
          data: [],
          isFetching: false,
        },
      },
    };
    wrapper = shallow(<AdvancedSearch {...props} />);
  });

  test('should render', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  test('should call to get search terms', () => {
    expect(props.dispatchFetchSearchTermsIfNeeded).toHaveBeenCalled();
  });

  test('should call to get evidences', () => {
    expect(props.dispatchfetchEvidencesIfNeeded).toHaveBeenCalledWith('go');
    expect(props.dispatchfetchEvidencesIfNeeded).toHaveBeenCalledWith('annotation');
  });

  test('should add field rows', () => {
    wrapper.find('#add-field').simulate('click');
    expect(props.dispatchAddClause).toHaveBeenCalled();
  });

  test('should submit a query', () => {
    wrapper.find('#submit-query').simulate('click');
    expect(props.handleAdvancedSubmitClick).toHaveBeenCalled();
  });
});
