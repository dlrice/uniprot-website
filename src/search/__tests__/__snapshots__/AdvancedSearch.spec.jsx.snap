// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdvancedSearch shallow components should render 1`] = `
"<div className=\\"advanced-search\\">
  <div>
    <label htmlFor=\\"namespace-select\\">
      Searching in
      <select id=\\"namespace-select\\">
        <option>
          UniProtKB
        </option>
      </select>
    </label>
  </div>
  <ClauseList dispatchAddClause={[Function: mockConstructor]} handleAdvancedSubmitClick={[Function: mockConstructor]} dispatchFetchSearchTerms={[Function: mockConstructor]} dispatchfetchEvidencesIfNeeded={[Function: mockConstructor]} dispatchCopyQueryClausesToSearch={[Function: mockConstructor]} history={{...}} clauses={{...}} namespace=\\"UniProtKB\\" searchTerms={{...}} evidences={{...}} />
  <button type=\\"button\\" id=\\"add-field\\" className=\\"button\\" onClick={[Function: mockConstructor]}>
    Add Field
  </button>
  <div>
    <button type=\\"button\\" id=\\"submit-query\\" className=\\"button\\" onClick={[Function: mockConstructor]}>
      Search
    </button>
  </div>
</div>"
`;
